examples/inspection/plot_model_inspection_cross_validation.py

```diff
@@ -0,0 +1,232 @@
+# and a set that we will leave out for further confirmation.
+
+# %%
+from sklearn.datasets import fetch_california_housing
+from sklearn.model_selection import train_test_split
+
+X, y = fetch_california_housing(as_frame=True, return_X_y=True)
+y -= y.mean()
+
+X_train, X_test, y_train, y_test = train_test_split(
+    X, y, test_size=0.1, random_state=0)
+
+# %%
+# Predictive model
+# ----------------
+#
+# In this example, we will use a linear model which should be a good baseline:
+# a :ref:`ridge model <ridge_regression>`. A ridge model enforces a L2 penalty
+# on the coefficients. The penalty parameter `alpha` thus has to be tuned. More
+# importantly, this parameter needs to be tuned for our specific problem:
+# tuning on another dataset does not ensure an optimal parameter value for the
+# current dataset.
+#
+# Here, we use the class :class:`~sklearn.linear_model.RidgeCV` that can tune
+# `alpha` by cross-validation.
+#
+# We also add a preprocessing stage to :ref:`standardize
+# <preprocessing_scaler>` the data such that the regularization strength is
+# applied homogeneously on each coefficient.
+
+# %%
+import numpy as np
+from sklearn.linear_model import RidgeCV
+from sklearn.pipeline import make_pipeline
+from sklearn.preprocessing import StandardScaler
+
+alphas = np.logspace(-1, 2.5, num=50)
+model = make_pipeline(StandardScaler(), RidgeCV(alphas=alphas))
+
+# %%
+# `model` is a machine learning pipeline made of the preprocessing stage
+# followed by a ridge regressor that will perform an internal cross-validation
+# at `fit` to select the best parameter `alpha`. The candidates for the
+# parameter `alpha` are given by the variable `alphas`.
+#
+# Cross-validation framework
+# --------------------------
+#
+# Before putting such a predictive model into production, one needs to evaluate
+# the performance of the model to have an idea of what to expect in production.
+#
+# Cross-validation should be used to make this analysis. First, it allows us to
+# quantify the variance of the model performance. A large variance of the score
+# metric will indicate that we cannot trust the reported performance nor try to
+# interpret findings built on internal model's parameters. Usually large
+# variations are linked to small sample size but not only.
+
+# %%
+from sklearn.model_selection import cross_validate
+from sklearn.model_selection import RepeatedKFold
+
+cv = RepeatedKFold(n_splits=10, n_repeats=10, random_state=0)
+cv_results = cross_validate(
+    model, X_train, y_train, cv=cv, return_estimator=True, n_jobs=2)
+
+# %%
+# Here, we used a repeated K-fold cross-validation. At each round of
+# cross-validation, it should be noted that the parameter `alpha` of the ridge
+# regressor is also optimized via another internal cross-validation. This
+# process is called a nested cross-validation and should always be implemented
+# whenever model's parameters need to be optimized.
```

Comment:
Nested cross validation takes much longer to train and I do not know if it is always recommended.

In this example, the nested cross validation only searches through one parameter and that parameter does not vary across folds. This leads to a model selection process leading to `alpha=40`.

In general, one can search through many parameter combinations, which would lead to different models. I.E, there would be one model for each loop of the outer fold. I do not think there is a way to decide which hyper parameter combination one should use in this case. (Unless you are ensembling them)

Reply:
> In general, one can search through many parameter combinations, which would lead to different models. I.E, there would be one model for each loop of the outer fold. I do not think there is a way to decide which hyper parameter combination one should use in this case. (Unless you are ensembling them)

I see your point here.

I still think this is beneficial to do the nested-cross-validation even with many hyper-parameters. The variability of the hyper-parameters would be informative. And I am not aware of any other way to make this analysis with a less costly strategy. However, this is true that it would not necessarily help you in choosing a specific configuration.

Do you have a proposal to mitigate the statement?

Reply:
For me, the wording is a bit absolute. One could state that it is recommended, but certainly not "should always be implemented".